multisig_wallet/transaction.py

```diff
@@ -97,48 +73,61 @@ def description(self) -> str:
+        params = "" if params is None else params
 
-        return Transaction(executed=transaction_executed,
-                           destination=Address.from_bytes(destination.strip(b'\x00')),
-                           method=method.decode(encoding="utf-8"),
-                           params=params.decode(encoding="utf-8"),
-                           value=int.from_bytes(value, DATA_BYTE_ORDER),
-                           description=description.decode(encoding="utf-8"))
+        if len(method) > MAX_METHOD_LEN \
+                or len(params) > MAX_PARAMS_LEN \
+                or len(description) > MAX_DESCRIPTION_LEN:
+            revert("too long parameter length")
+        try:
+            value.to_bytes(DEFAULT_VALUE_BYTES, DATA_BYTE_ORDER)
+        except OverflowError:
+            revert("exceed ICX amount you can send at one time")
+
+        return cls(executed=executed,
+                   destination=destination,
+                   value=value,
+                   method=method,
+                   params=params,
+                   description=description)
+
+    @classmethod
+    def from_bytes(cls, buf: bytes):
+        encoded_executed = buf[0]
+        encoded_destination = buf[1: 1 + ADDRESS_BYTE_LEN]
+        encoded_value = buf[1 + ADDRESS_BYTE_LEN: 1 + ADDRESS_BYTE_LEN + DEFAULT_VALUE_BYTES]
+        flexible_vars_json_string = buf[1 + ADDRESS_BYTE_LEN + DEFAULT_VALUE_BYTES:].decode()
+        flexible_vars_json = json_loads(flexible_vars_json_string)
+
+        return cls(executed=encoded_executed,
+                   destination=Address.from_bytes(encoded_destination.strip(b'\x00')),
+                   value=int.from_bytes(encoded_value, DATA_BYTE_ORDER),
+                   method=flexible_vars_json["method"],
+                   params=flexible_vars_json["params"],
+                   description=flexible_vars_json["description"])
 
     def to_bytes(self) -> bytes:
-        flexible_struct_format_bytes = self._flexible_struct_format.encode()
-        flexible_struct_format_len = len(flexible_struct_format_bytes)
-        if flexible_struct_format_len > 255:
-            revert("too long parameters")
-
-        packed_fixed_variables = pack(
-            Transaction._fixed_struct_format,
-            self.executed,
-            self.destination.to_bytes(),
-            self._value.to_bytes(DEFAULT_VALUE_BYTES, DATA_BYTE_ORDER),
-            flexible_struct_format_len
-        )
-
-        packed_flexible_variables = pack(
-            self._flexible_struct_format,
-            self._method.encode(encoding="utf-8"),
-            self._params.encode(encoding="utf-8"),
-            self._description.encode(encoding="utf-8"))
-
-        transaction = packed_fixed_variables + \
-                      flexible_struct_format_bytes + \
-                      packed_flexible_variables
-
-        return transaction
+        encoded_executed = self.executed.to_bytes(1, DATA_BYTE_ORDER)
```

Comment:
you should consider negative number.
because negative value is added 1 byte than positive number

Reply:
I have added code calling '_only_positive_number' in submitTransaction method.